### DISCUSSION TITLE
[Proposal]: await using and ConfigureAwait without going back to pre C# 8 indentation

### DISCUSSION THREAD

**Discussion by virzak** [dotnet/csharplang#4134] [Language Ideas] [answered]
# [Proposal]: await using and ConfigureAwait without going back to pre C# 8 indentation

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Allow C# 8 using declaration with await using  and .ConfigureAwait(false)

Current syntax:
```csharp
var c = new MyAsyncDisposableClass();
await using (c.ConfigureAwait(false))
{
    ...
}
```

Proposed syntax:

```csharp
var c = new MyAsyncDisposableClass();
await using c.ConfigureAwait(false);
...
```

## Motivation
[motivation]: #motivation

With release of C# 8, most `using` blocks were converted to using declarations. It was great for removing unnecessary  indentation.

`IAsyncDisposable` a less advertised feature and didn't prominently feature in samples until later. Here is a recent [commit](https://github.com/dotnet/EntityFramework.Docs/commit/31ebfae2797a045358bb52a31e5f97c35a377c06#diff-74adba6e8a4dfeded2add364ca3a4c03f389389660c6f6a7313ec038523c4ad2R118) where EF Core samples changed to `await using`.

Our project, which creates a lot of `DbContext` and `FileStream` instances benefitted from reducing indentation. The using statement usually  occurs once per method. [ConfigureAwait FAQ](https://devblogs.microsoft.com/dotnet/configureawait-faq/#can-i-use-configureawait-when-await-using-an-iasyncdisposable)'s example of `.ConfigureAwait` [extension](https://github.com/dotnet/runtime/blob/ce1d60ff19d76506eca59d21af47151d0fceb55c/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskAsyncEnumerableExtensions.cs#L16) on IAsyncDisposable takes the project back to the old indentation. 

## Detailed design
[design]: #detailed-design

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

Not sure how important it is to use the ConfigureAwait extension, since the EF Core sample doesn't use it and there is no compiler / analyzer warning if omitted.

**Comment by svick** [comment, marked as answer]
Note that you can write the code like this:

```c#
await using var _ = c.ConfigureAwait(false);
```

(This actually declares a variable named `_`, it's not a discard.)

Also, I think this is pretty much a duplicate of https://github.com/dotnet/csharplang/issues/2235.

**Reply by virzak** [reply]
@svick, yep that works.
